sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetUtils.scala

```diff
@@ -420,6 +420,22 @@ object ParquetUtils extends Logging {
     statistics.getNumNulls;
   }
 
+  // Replaces each VariantType in the schema with the corresponding type in the shredding schema.
+  // Used for testing, where we force a single shredding schema for all Variant fields.
+  // Does not touch Variant fields nested in arrays, maps, or UDTs.
+  private def replaceVariantTypes(schema: StructType, shreddingSchema: StructType): StructType = {
+    val newFields = schema.fields.zip(shreddingSchema.fields).map {
+      case (field, shreddingField) =>
+        field.dataType match {
+          case s: StructType =>
+            field.copy(dataType = replaceVariantTypes(s, shreddingSchema))
```

Comment:
Have you considered adding a property like `shreddedSchema` in `VariantType`? This way, we won't need to determine how this struct type was from. Currently, it seems that this is judged through __VARIANT_WRITE_SHREDDING_KEY."

Reply:
Hi @Zouxxyy, that's actually what I tried initially. I could go back to that approach if there's a consensus that it's better, but I felt that it had some downsides. Specifically, the approach I took was to add an optional `shreddingSchema` property to the VariantType class, but leave the `case object VariantType` to represent a VariantType with no shredding (similar to what's been done for StringType with default collation), to avoid having to modify all of the existing code that pattern matches to `VariantType`.

I didn't think this approach added much value - it's not very clear what it means for the type to have a shredding schema outside of the specific context of the couple of writer classes that care about it, and it felt like it was just pushing writer-specific code into the core of the type system. E.g. when serializing the type to a string, should we include the shredding schema? I think the answer is that we'd want to when sending it to ParquetWriteSupport, but not when writing it to the parquet file (since it's not part of the Spark type when reading it back from Parquet). It seemed better to resolve these ambiguities explicitly in the relevant writer code, rather than making them part of the data type definition.

Reply:
@cashmand Thank you for your reply. We have also implemented the shredding logic internally in Apache Paimon. Specifically, we use `VariantType` along with an optional `ShreddSchema`. In particular, on the write side, we calculate the `ShreddSchema` based on the table's shredding props and then set it into `VariantType`, which is passed to our writer (currently Parquet, but it could be ORC in the future, etc.).

On the read side, we calculate the required schema based on the scan filter and projection, which is also written into the optional attribute of VariantType, and passed to our reader. Combined with the schema of the Parquet variant, we generate the trimmed read ShreddSchema. I think that shredding is an extension of the variant type; when our writer sees, oh, there exists a ShreddSchema, the writer immediately knows that shredding needs to be performed.

Reply:
shredding only matters for read/write and I think it's weird to put the info in the `VariantType` which can also be consumed by SQL functions/operators. The current approach looks cleaner to me.

Reply:
Hi @Zouxxyy, I tend to agree with @cloud-fan's comment. My preference would be to stick with the current approach for now. I think it should be entirely contained within the Parquet writer code, so if we decide to extend the VariantType later, I don't think it would be hard to change this part of the code to use a different approach.

Reply:
Got it, It does look much more concise for now, looking forward to its completion!